tests/training_success/explore_schema.py

```diff
@@ -0,0 +1,422 @@
+    if "maximum" in item:
+        maximum = item["maximum"]
+        candidates = [num for num in candidates if num < maximum]
+    return candidates + [random.random() * 0.99 * maximum]
+
+
+def explore_integer(item: Dict[str, Any]) -> List[ParameterBaseTypes]:
+    """Return possible parameter values for the `integer` parameter type.
+
+    Args:
+        item: dictionary containing details on the parameter such as default, min and max values.
+    TODO(Wael): Improve logic.
+    """
+    minimum, maximum = 0, 10
+
+    if "default" not in item or item["default"] is None:
+        candidates = []
+    else:
+        candidates = [item["default"], 2 * (item["default"] + 1), item["default"] // 2, -1 * item["default"]]
+
+    if "minimum" in item:
+        minimum = item["minimum"]
+        candidates = [num for num in candidates if num >= item["minimum"]]
+    if "maximum" in item:
+        maximum = item["maximum"]
+        candidates = [num for num in candidates if num <= item["maximum"]]
+
+    return candidates + [random.randint(minimum, maximum)]
+
+
+def explore_string(item: Dict[str, Any]) -> List[ParameterBaseTypes]:
+    """Return possible parameter values for the `string` parameter type.
+
+    Args:
+        item: dictionary containing details on the parameter such as default, min and max values.
+    """
+
+    if "enum" in item:
+        return item["enum"]
+    return [item["default"]]
+
+
+def explore_boolean() -> List[bool]:
+    """Return possible parameter values for the `boolean` parameter type (i.e. [True, False])"""
+    return [True, False]
+
+
+def explore_null() -> List[None]:
+    """Return possible parameter values for the `null` parameter type (i.e. [None])"""
+    return [None]
+
+
+def generate_possible_configs(config_options: Dict[str, Any]):
+    """Generate exhaustive configs from config_options.
+
+    This function does not take a cross product of all the options for all the config parameters. It selects parameter
+    values independently from each other.
+
+    Args:
+        config_options: dictionary mapping from ludwig config parameter to all values to be explored.
+            Here's an example of what it could look like:
+
+                trainer.batch_size: ["auto", 2, 43]
+                trainer.learning_rate: ["auto", 0.1, 0.00002, 0.32424]
+                ...
+    """
+    num_configs = 1
+    for key in config_options:
+        if isinstance(config_options[key], list):
+            num_configs = max(num_configs, len(config_options[key]))
+            config_options[key] = deque(config_options[key])
+
+    for _ in range(num_configs):
+        config = {}
+        for key in config_options:
+            if config_options[key] and not isinstance(config_options[key], str):
+                config[key] = config_options[key].popleft()
+            elif isinstance(config_options[key], str):
+                config[key] = config_options[key]
+        yield config
```

Comment:
nit/suggestion: It seems potentially cleaner to simply `yield create_nested_dict(config)` since AFAICT you're not using the intermediate config dict with absolute parameter path keys.